src/directory/s3_directory.rs

```diff
@@ -0,0 +1,285 @@
+use common::make_io_err;
+use directory::Directory;
+use directory::error::{IOError, OpenWriteError, OpenReadError, DeleteError, OpenDirectoryError};
+use directory::ReadOnlySource;
+use directory::shared_vec_slice::SharedVecSlice;
+use directory::WritePtr;
+use std::collections::HashMap;
+use std::convert::From;
+use std::default::Default;
+use std::error::Error;
+use std::fmt;
+use std::io;
+use std::path::{Path, PathBuf};
+use std::result;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::sync::RwLock;
+use rusoto_core::{DefaultCredentialsProvider, Region, default_tls_client};
+use rusoto_s3::{S3, S3Client, HeadBucketRequest, GetObjectRequest};
+
+fn get_client(region: Region) -> Result<Box<S3>, Box<Error>> {
+    // TODO: handle missing creds
+    let client = default_tls_client()?;
+    let provider = DefaultCredentialsProvider::new()?;
+
+    Ok(Box::new(S3Client::new(client, provider, region)))
+}
+
+#[derive(Clone)]
+struct InnerDirectory {
+    bucket: String,
+    cache: Arc<RwLock<HashMap<PathBuf, Arc<Vec<u8>>>>>,
+}
+
+impl InnerDirectory {
+    fn new(bucket: String) -> InnerDirectory {
+        InnerDirectory {
+            cache: Arc::new(RwLock::new(HashMap::new())),
+            bucket,
+        }
+    }
+
+    fn load_missing_key(&self, client: &S3, path: &Path) -> Result<(), OpenReadError> {
+        let mut map = self.cache.write().map_err(|_| {
+            let msg = format!(
+                "Failed to acquire write lock for the \
+                                            directory when trying to read {:?}",
+                path
+            );
+            let io_err = make_io_err(msg);
+            OpenReadError::IOError(IOError::with_path(path.to_owned(), io_err))
+        })?;
+
+        // TODO: this is comical and I'm more than likely over thinking it
+        let key = path.as_os_str().to_os_string().into_string().map_err(|_| {
+            let msg = format!("Could not build key path");
+            let io_err = make_io_err(msg);
+            OpenReadError::IOError(IOError::with_path(path.to_owned(), io_err))
+        })?;
+
+        let obj = client
+            .get_object(&GetObjectRequest {
+                bucket: self.bucket.clone(),
+                key,
+                ..Default::default()
```

Comment:
The `rusoto` api has a large number of property fields, and each `rusoto` request object implements the default trait. This way you can just specify the 1 or 2 fields that you need then apply the `..` operator + `Default::default` to set all the other fields to their default value. Often this is `None` in `rusoto`. Saved me a lot of boiler plate typing for sure.